pkg/controller/servicebindingrequest/reconciler.go

```diff
@@ -135,49 +118,62 @@ func (r *Reconciler) Reconcile(request reconcile.Request) (reconcile.Result, err
 	logger = logger.WithValues("ServiceBindingRequest.Name", sbr.Name)
 	logger.Debug("Found service binding request to inspect")
 
-	// splitting instance from it's status
-	sbrStatus := &sbr.Status
+	ctx := context.Background()
+
+	selectors := extractServiceSelectors(sbr)
+	if len(selectors) == 0 {
+		v1.SetStatusCondition(&sbr.Status.Conditions, v1.Condition{
+			Type:    conditions.BindingReady,
+			Status:  corev1.ConditionFalse,
+			Reason:  conditions.EmptyServiceSelectorsReason,
+			Message: "The spec.backingServiceSelectors field is empty.",
+		})
+		_, updateErr := updateServiceBindingRequestStatus(r.dynClient, sbr)
+		if updateErr == nil {
+			return Done()
+		}
+		return NoRequeue(ErrEmptyBackingServiceSelectors)
+	}
+
+	serviceCtxs, err := buildServiceContexts(r.dynClient, sbr.GetNamespace(), selectors, nil)
+	if err != nil {
+		return RequeueError(err)
+	}
+
+	binding, err := buildBinding(
+		r.dynClient,
+		sbr.Spec.CustomEnvVar,
+		serviceCtxs,
+		sbr.Spec.EnvVarPrefix,
+	)
+	if err != nil {
+		return RequeueError(err)
+	}
 
 	options := &ServiceBinderOptions{
 		Client:                 r.client,
 		DynClient:              r.dynClient,
 		DetectBindingResources: sbr.Spec.DetectBindingResources,
-		EnvVarPrefix:           sbr.Spec.EnvVarPrefix,
 		SBR:                    sbr,
 		Logger:                 logger,
+		Objects:                serviceCtxs.GetObjects(),
+		Binding:                binding,
 	}
 
-	bm, err := BuildServiceBinder(options)
+	sb, err := BuildServiceBinder(ctx, options)
 	if err != nil {
-		logger.Error(err, "Creating binding context")
-		if err == EmptyBackingServiceSelectorsErr || err == EmptyApplicationSelectorErr {
-			// TODO: find or create an error type containing suitable information to be propagated
-			var reason string
-			if errors.Is(err, EmptyBackingServiceSelectorsErr) {
-				reason = "EmptyBackingServiceSelectors"
-			} else {
-				reason = "EmptyApplicationSelector"
-			}
-
-			v1.SetStatusCondition(&sbr.Status.Conditions, v1.Condition{
-				Type:    conditions.BindingReady,
-				Status:  corev1.ConditionFalse,
-				Reason:  reason,
-				Message: err.Error(),
-			})
-			_, updateErr := updateServiceBindingRequestStatus(r.dynClient, sbr)
-			if updateErr == nil {
-				return Done()
-			}
-		}
-		return RequeueError(err)
+		// BuildServiceBinder can return only InvalidOptionsErr, and it is a programmer's error so
+		// just bail out without re-queueing nor updating conditions.
+		logger.Error(err, "Building ServiceBinder")
```

Comment:
Please improve the log message.

Reply:
Do you have a suggestion?

From our current usage, the error message is used to give context and the error is added in the structured log output.

The only difference from other logger usages is that "On" was dropped.

Reply:
Either move the error logging into `BuildServiceBinder` and there you can provide more infos what options are invalid, or log here what you have written in the comment above:

```
logger.Error(err, "No requeueing nor updating conditions")
```

Reply:
`BuildServiceBinder` single purpose is to guarantee the options are valid to construct a complete `ServiceBinder`. 

It already returns an error that can be improved to better report the reaason options might be invalid.

Reply:
Solved in 865cca87031795649d6e02ee0160fc58841334b0.